### PR TITLE
Fix crash of tf.image.crop_and_resize when input is large number

### PR DESCRIPTION
**Pull request by yongtang**
This PR is part of the effort in #46890 where
tf.image.crop_and_resize will crash if shape consists of large number.

Signed-off-by: Yong Tang <yong.tang.github@outlook.com>

### CODE REVIEW

**Review by mihaimaruseac — APPROVED** [review]

**Comment by gbaned** [comment]
Seems auto-merge is not happening but the changes are merged into master now, so we can close this. Thank you for the PR.

**Comment by joker-eph** [comment]
The test provided in this PR breaks with the sanitizers:

```
tensorflow/core/grappler/costs/op_level_cost_estimator.cc:1547:17: runtime error: signed integer overflow: 22719123801 * 1145309325 cannot be represented in type 'long'
    #0 0x562c0294ff1f in tensorflow::grappler::OpLevelCostEstimator::CalculateTensorElementCount(tensorflow::OpInfo_TensorProperties const&, bool*) tensorflow/core/grappler/costs/op_level_cost_estimator.cc:1547:17
    #1 0x562c0294a581 in tensorflow::grappler::OpLevelCostEstimator::PredictCropAndResize(tensorflow::grappler::OpContext const&, tensorflow::grappler::NodeCosts*) const tensorflow/core/grappler/costs/op_level_cost_estimator.cc:2643:35
    #2 0x562c0296349e in operator() tensorflow/core/grappler/costs/op_level_cost_estimator.cc:365:14
    #3 0x562c0296349e in __invoke<(lambda at tensorflow/core/grappler/costs/op_level_cost_estimator.cc:364:12) &, const tensorflow::grappler::OpContext &, tensorflow::grappler::NodeCosts *> include/c++/v1/type_traits:3918:1
    #4 0x562c0296349e in __call<(lambda at tensorflow/core/grappler/costs/op_level_cost_estimator.cc:364:12) &, const tensorflow::grappler::OpContext &, tensorflow::grappler::NodeCosts *> include/c++/v1/__functional/invoke.h:30:16
    #5 0x562c0296349e in operator() include/c++/v1/__functional/function.h:221:12
    #6 0x562c0296349e in tensorflow::Status std::__u::__function::__policy_invoker<tensorflow::Status (tensorflow::grappler::OpContext const&, tensorflow::grappler::NodeCosts*)>::__call_impl<std::__u::__function::__default_alloc_func<tensorflow::grappler::OpLevelCostEstimator::OpLevelCostEstimator()::$_2::operator()(tensorflow::Status (tensorflow::grappler::OpLevelCostEstimator::*)(tensorflow::grappler::OpContext const&, tensorflow::grappler::NodeCosts*) const) const::'lambda'(tensorflow::grappler::OpContext const&, tensorflow::grappler::NodeCosts*), tensorflow::Status (tensorflow::grappler::OpContext const&, tensorflow::grappler::NodeCosts*)> >(std::__u::__function::__policy_storage const*, tensorflow::grappler::OpContext const&, tensorflow::grappler::NodeCosts*) include/c++/v1/__functional/function.h:702:16
    #7 0x562c0294cd22 in operator() include/c++/v1/__functional/function.h:834:16
    #8 0x562c0294cd22 in operator() include/c++/v1/__functional/function.h:1175:12
    #9 0x562c0294cd22 in tensorflow::grappler::OpLevelCostEstimator::PredictNodeCosts(tensorflow::grappler::OpContext const&, tensorflow::grappler::NodeCosts*) const tensorflow/core/grappler/costs/op_level_cost_estimator.cc:702:12
    #10 0x562c0294b58d in tensorflow::grappler::OpLevelCostEstimator::PredictCosts(tensorflow::grappler::OpContext const&) const tensorflow/core/grappler/costs/op_level_cost_estimator.cc:653:7
    #11 0x562c0292c112 in tensorflow::grappler::AnalyticalCostEstimator::PredictCosts(tensorflow::GraphDef const&, tensorflow::RunMetadata*, tensorflow::grappler::Costs*) const tensorflow/core/grappler/costs/analytical_cost_estimator.cc:201:35
    #12 0x562c029277c7 in tensorflow::grappler::VirtualCluster::Run(tensorflow::grappler::GrapplerItem const&, tensorflow::RunMetadata*) tensorflow/core/grappler/clusters/virtual_cluster.cc:88:3
    #13 0x562c027cbdfc in tensorflow::grappler::GraphMemory::InferStatically(std::__u::unordered_map<std::__u::basic_string<char, std::__u::char_traits<char>, std::__u::allocator<char> >, tensorflow::DeviceProperties, std::__u::hash<std::__u::basic_string<char, std::__u::char_traits<char>, std::__u::allocator<char> > >, std::__u::equal_to<std::__u::basic_string<char, std::__u::char_traits<char>, std::__u::allocator<char> > >, std::__u::allocator<std::__u::pair<std::__u::basic_string<char, std::__u::char_traits<char>, std::__u::allocator<char> > const, tensorflow::DeviceProperties> > > const&) tensorflow/core/grappler/costs/graph_memory.cc:40:22
    #14 0x562c0279dcc3 in IdentifySwappingCandidates tensorflow/core/grappler/optimizers/memory_optimizer.cc:987:31
```

can you take a look?

File: tensorflow/python/ops/image_ops_test.py [hunk 1]
```diff
@@ -6075,6 +6075,16 @@ def testImageCropAndResize(self):
...
         self.evaluate(op)
 
+  def testImageCropAndResizeWithInvalidInput(self):
+    with self.session():
+      with self.assertRaises((errors.InternalError, ValueError)):
+        op = image_ops_impl.crop_and_resize_v2(
+            image=np.ones((1, 1, 1, 1)),
+            boxes=np.ones((11, 4)),
+            box_indices=np.ones((11)),
+            crop_size=[2065374891, 1145309325])
```
Comment (joker-eph) [inline tensorflow/python/ops/image_ops_test.py:6085]: The issue is that this produces a tensor with a shape that describes a number of elements that can't be represented in an int64_t, do we need such a large size here? Surely we can't execute anything like this right?

**Comment by yongtang** [comment]
Thanks @joker-eph for the comment. In the test, the large number was used on purpose so that tensorflow can trigger an error gracefully (as was shown in the test setup):
https://github.com/tensorflow/tensorflow/blob/c8d87055a56d8740d27ad8bdc74a7459ede6900e/tensorflow/python/ops/image_ops_test.py#L6080

Originally, the above test will cause tensorflow to crash with undefined behavior, which may potentially be a security issue.

So the test itself is expected.

From the sanitizer error in logs:
```
tensorflow/core/grappler/costs/op_level_cost_estimator.cc:1547:17: runtime error: signed integer overflow: 22719123801 * 1145309325 cannot be represented in type 'long'
    #0 0x562c0294ff1f in tensorflow::grappler::OpLevelCostEstimator::CalculateTensorElementCount(tensorflow::OpInfo_TensorProperties const&, bool*) tensorflow/core/grappler/costs/op_level_cost_estimator.cc:1547:17
    #1 0x562c0294a581 in tensorflow::grappler::OpLevelCostEstimator::PredictCropAndResize(tensorflow::grappler::OpContext const&, tensorflow::grappler::NodeCosts*) const tensorflow/core/grappler/costs/op_level_cost_estimator.cc:2643:35
    #2 0x562c0296349e in operator() tensorflow/core/grappler/costs/op_level_cost_estimator.cc:365:14
```

I think the issue is actually coming from https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/grappler/costs/op_level_cost_estimator.cc#L1547

which should result in an status error (when multiply overflows) and be populated to `Status OpLevelCostEstimator::PredictCropAndResize` instead:
https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/grappler/costs/op_level_cost_estimator.cc#L2621

Let me see if I can create a PR to address the issue.

**Comment by joker-eph** [comment]
I think there is a more fundamental question: should we allow this in the first place in the graph? What is the use case?
I'm worried of chasing this everywhere in the codebase instead of forbidding such shape from being created entirely for example.

**Comment by joker-eph** [comment]
For example, the fix implemented in this PR in crop_and_resize_op.cc using AddDimWithStatus seems in line with this: `AddDimWithStatus` exists because it seems invalid to create a shape when we can't compute the number of elements.

**Comment by yongtang** [comment]
@joker-eph Since `crop_size` is passed as input (not attr) of the node in the graph, it may only be checked when graph runs in case `crop_size` itself is a non-constant tensor I think.

On the other hand, it might be possible to do some preliminary check so that in case crop_size is a constant, then the graph can return an error before it runs. 

In tensorflow 1.x, the shape is pre-checked in the following (before graph runs):
https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/ops/image_ops.cc#L879-L908
so it will be possible to do some preliminary check in case `crop_size` itself is a constant tensor.

However, in tensorflow 2.x's eager mode, the path to do the shape check (and resolve and constant tensor) before graph runs does not go through the above path any more. As a result the above function will not be triggered in eager mode anymore. As such a potential shape check is not possible like 1.x.

**Comment by joker-eph** [comment]
I meant that we could forbid at the NodeDef creation to create such shapes, so that it would never appear in the Graph. I'm not sure about all the implication, I just doubt the stack is robust on this.

**Pull request closed without merging**